[CDAP-9121] Authorization cache redesign

Nishith opened the pull request:
- Removed the service that ran on Master and updated its cache by calling the authorizer’s listPrivilege
- The cache on the containers is now populated on an enforce call and the result of the enforcement is cached.
- Change the Authorization cache key to [Principal, Entity, Action]
- Remove the Proxy service that the containers used to refresh their caches
- Authorization cache is now limited by the number of entries it can hold
- Containers now call Master to enforce authorization policies and then cache the results

Based on the design documented at https://wiki.cask.co/pages/viewpage.action?pageId=4363553

Commit:
Added a configuration parameter to enable or disable hierarchical privileges.

Commit:
Authorization redesign

Commit:
WIP

Commit:
Created an abstract class for privilege enforcers

Commit:
Added unit test for RemoteAuthorizationEnforcer

Commit:
Fixed a privilege propagation bug in RemoteAuthorizationEnforcer

Commit:
Fixed tests

Nishith commented:
Build running here: https://builds.cask.co/browse/CDAP-DUT5690-19

anwar6953 commented on cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/AbstractAuthorizer.java:
```diff
@@ -62,16 +61,16 @@ public void enforce(EntityId entity, Principal principal, Action action) throws
+        try {
+          enforce(entityId, principal, new HashSet<>(Arrays.asList(Action.values())));
+          return true;
+        } catch (Exception e) {
```
Shouldn't you only return `false` if there's a `UnauthorizedException`?
If there's other types of Exceptions, don't we want to surface those, instead of being silent about the exception?

Nishith replied:
It can only throw either an UnauthorizedException or an ExecutionException from the loading cache's get which will be thrown if the load method throws an exception, the load method here can only throw UnauthorizedException. So, the behavior is correct but I agree it is not apparent. I will add a check to improve both safety and clarity.

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,94 @@
+    for (Action action : actions) {
+      try {
+        enforce(entity, principal, action);
+      } catch (Exception ex) {
```
same here

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,94 @@
+        disallowed.add(action);
+      }
+    }
+    if (disallowed.size() > 0) {
```
why not fail fast? Why check other actions, if the first action is disallowed?

Nishith replied:
We want to tell the user which privileges they are missing, makes it easier to correct it.

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,94 @@
+          try {
+            enforce(entityId, principal, action);
+            return true;
+          } catch (Exception ignored) {
```
same here - only ignore `UnauthorizedException`s?

Commit:
Addressed comments

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java:
```diff
@@ -49,15 +46,25 @@
+  }
+
+  @POST
+  @Path("/enforce")
```
Who will call this?

Nishith replied:
This is called by the RemoteAuthorizationEnforcer.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2712,6 +2712,15 @@
+    <value>10000</value>
+    <description>
+      Number of entries to hold in the container authorization cache. This setting only takes effect if
+      ${security.authorization.cache.enabled} is set to true.
```
Why need two separate setting? Can't we just use this one and if it is 0, that means no cache.

chtyim commented on cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/AbstractAuthorizer.java:
```diff
@@ -62,16 +65,18 @@ public void enforce(EntityId entity, Principal principal, Action action) throws
       public boolean apply(EntityId entityId) {
-        return allowedEntities.contains(entityId);
+        try {
+          enforce(entityId, principal, new HashSet<>(Arrays.asList(Action.values())));
```
The set can be constant. Also you can simply use `EnumSet.allOf`

rohitsinha54 replied:
do this
```
EnumSet.allOf(Action.class)
```

chtyim commented on cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authorization/AbstractAuthorizer.java:
```diff
@@ -62,16 +65,18 @@ public void enforce(EntityId entity, Principal principal, Action action) throws
-    for (Privilege privilege : privileges) {
-      allowedEntities.add(privilege.getEntity());
-    }
+  public Predicate<EntityId> createFilter(final Principal principal) throws Exception {
```
Seems like this method should not have `throws Exception`, since it just create a filter.

Nishith replied:
I agree but that would be an interface change, ok to do?

chtyim replied:
Who implement this?

Nishith replied:
It is part of the Authorizer interface so could be implemented by extensions.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,96 @@
+    if (!isSecurityAuthorizationEnabled()) {
+      return;
+    }
+    Set<Action> disallowed = new HashSet<>(actions.size());
```
Use `EnumSet` instead.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,96 @@
+        disallowed.add(action);
+      }
+    }
+    if (disallowed.size() > 0) {
```
Use `!isEmpty` instead.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+    this.remoteClient = new RemoteClient(discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false), "/v1/execute/");
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    int cacheTtlSecs = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
```
`TTL`, should be all caps for acronym

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<AuthorizationPrivilege, Boolean>() {
+        @SuppressWarnings("NullableProblems")
```
What is this suppress for??

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Boolean load(AuthorizationPrivilege authorizationPrivilege) throws Exception {
+          LOG.trace("Cache miss for " + authorizationPrivilege);
```
Use `{}` instead of `+`.

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java:
```diff
@@ -49,15 +46,25 @@
+  public void enforce(HttpRequest request, HttpResponder responder) throws Exception {
+    Iterator<MethodArgument> arguments = parseArguments(request);
+    EntityId entityId = deserializeNext(arguments);
+    Principal principal = deserializeNext(arguments);
```
AuthorizationEnforcer has an API which takes Set<Action> it is possible that a client call that to save rpc call. We should make this handle both i think.

Nishith replied:
The key for the cache is Principal, Entity, and Action, so we need to enforce on individual actions. The reason behind this is that in the case of an operation requiring multiple privileges we should be able to tell the user which privilege they are missing. This method is called by the cache's load method in the case of a cache miss and that only has access to the key class, so this enforce will not be called with a Set. The reason to have two classes is that we don't want to force the user to grant and revoke on single actions and continue supporting the multi-action commands but we need to call enforce Action be Action.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+
+  private boolean doEnforce(AuthorizationPrivilege authorizationPrivilege) throws IOException {
+    HttpResponse response = executeRequest(authorizationPrivilege);
+    return HttpResponseStatus.OK.getCode() == response.getResponseCode();
```
Use `HttpURLConnection.HTTP_OK` from JDK instead.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+    return remoteClient.execute(request);
+  }
+
+  private static List<MethodArgument> createBody(Object... arguments) {
```
I am a bit confused by this method. How's the POST body looks like?? Shouldn't it be just json of the `AuthorizationPrivilege` object which should be shared between server and client?

rohitsinha54 replied:
Why do we need this ? Can't we just serialize/deserialize AuthorizationPrivilege ?

chtyim reviewed (commented):
Took a quick first pass

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java:
```diff
@@ -78,9 +85,9 @@ public void grant(HttpRequest request, HttpResponder responder) throws Exception
-    LOG.debug("Granted {} on {} to {} successfully", actions, entityId, principal);
+    LOG.debug("Granting {} on {} to {}", actions, entityId, principal);
+    privilegesManager.grant(entityId, principal, actions);
+    LOG.info("Granted {} on {} to {} successfully", actions, entityId, principal);
```
Not sure if its a good idea to make this and other below as info since this will lead to lot of log lines in master logs. Something which might not be very useful given a user can always do list to see the privilege. Should have this as debug like earlier.

Nishith replied:
I think someone getting a new privilege or losing a privilege is important information for the user to know. Also, this should not happen very frequently. Right now it will only happen on entity creation, deletion or on manual grants and revokes.

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppFabricServer.java:
```diff
@@ -162,9 +158,7 @@ protected void startUp() throws Exception {
-        pluginService.start(),
-        privilegesFetcherProxyService.start()
-      )
+        pluginService.start())
```
nit: indentation the last closing braces should be in the line below.

rohitsinha54 commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2712,6 +2712,15 @@
 
   <property>
+    <name>security.authorization.cache.max.entries</name>
+    <value>10000</value>
```
is there some reason to have it 10k and not something else or just a rough estimate ?

Nishith replied:
This is based on a single principal owning 2500 entities for user containers. Unfortunately, the cache does not have a size based limit.

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,96 @@
+ */
+public abstract class AbstractAuthorizationEnforcer implements AuthorizationEnforcer {
+
+  static final Predicate<EntityId> ALLOW_ALL = new Predicate<EntityId>() {
```
this can be private

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizerInstantiator.java:
```diff
@@ -107,12 +107,16 @@ public synchronized Authorizer get() {
 
     if (!authorizationEnabled) {
-      LOG.debug("Authorization is disabled. Using a no-op authorizer.");
+      LOG.info("Authorization is disabled. Authorization can be enabled  by setting " +
```
why change these to info ?

Nishith replied:
Changed the first one back to debug. The second one under `if (!authenticationEnabled) {` should be info because that is an incorrect configuration and user should be told how to fix it.

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,90 @@
+
+  private boolean doEnforce(EntityId entity, Principal principal,
+                            Set<Action> actions, boolean exceptionOnFailure) throws Exception {
+    // If privilege propagation is enabled then check for the privilege on the parent, if any.
```
This recursion seems a bit weird. It walk up the hierarchy first if propagation is enabled and then does the check. Would be more better to enforce on the entity in question and if privilege exist on the entity we don't need to walk up the hierarchy if it doesn't then we do.

Nishith replied:
This was to retain the current behavior when privilege propagation is enabled but I agree this is wierd, I have changed it to search bottom up.

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,90 @@
+    return true;
+  }
+
+  private boolean isPrivilegePropagationEnabled() {
```
i dont think we should have a private getter for a private member  variable. Just use the variable directly.

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultPrivilegesManager.java:
```diff
@@ -19,75 +19,43 @@
 /**
  * A {@link PrivilegesManager} that also invalidates privileges caches when privileges are updated.
  */
 public class DefaultPrivilegesManager implements PrivilegesManager {
```
why do we need this class ?

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+    HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "enforce")
+      .withBody(GSON.toJson(createBody(authorizationPrivilege.getEntityId(),
+                                       authorizationPrivilege.getPrincipal(),
+                                       authorizationPrivilege.getAction())))
```
In earlier comment I raised the need for having endpoint handling Set<Action> so that we make one call. It seems like AuthorizationPrivilege only have action. Also if I remember correctly so far we havent seen an enforce call with actions rather than action so might not be even needed.

Nishith replied:
I don't understand the last line, are you saying that we haven't seen an enforce call with multiple actions so we don't need to handle Set?

rohitsinha54 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+   * Key for caching Privileges on containers. This represents a specific privilege on which authorization can be
+   * enforced. The cache stores whether the enforce succeeded or failed.
+   */
+  public static class AuthorizationPrivilege {
```
We should use AuthorizationRequest rather than adding new one or the other way

Nishith replied:
Not sure how to do that. We want to cache privileges for individual actions but we also want to let people grant and revoke on multiple actions.

rohitsinha54 commented on cdap-security/src/test/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcerTest.java:
```diff
@@ -0,0 +1,255 @@
+      DatasetId ds = NS.dataset("ds");
+      // All enforcement operations should succeed, since authorization is disabled
+      authEnforcementService.enforce(NS, ALICE, Action.ADMIN);
+      authEnforcementService.enforce(ds, BOB, Action.ADMIN);
```
what happens if someone does grant/revoke calls in disabled ?

rohitsinha54 commented on cdap-security/src/test/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcerTest.java:
```diff
@@ -0,0 +1,255 @@
+      try {
+        authorizationEnforcer.enforce(APP, ALICE, Action.ADMIN);
+        Assert.fail("Alice should not have ADMIN privilege on the APP.");
+      } catch (Exception ignored) {
```
this should be UnauthorizedException not just Exception ?

chtyim commented:
Also, please resolve conflict.

rohitsinha54 commented on cdap-security/src/test/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcerTest.java:
```diff
@@ -0,0 +1,255 @@
+      // because the LoadingCache should make a blocking call to retrieve his privileges
+      authEnforcementService.enforce(ds, BOB, Action.ADMIN);
+      // revoke all of alice's privileges
+      authorizer.revoke(NS);
```
should also test with revoke one action

rohitsinha54 commented on cdap-security/src/test/java/co/cask/cdap/security/authorization/AuthorizationCacheInvalidationTest.java:
```diff
@@ -1,146 +0,0 @@
- * Tests invalidation of caches in {@link DefaultAuthorizationEnforcementService} and
- * {@link PrivilegesFetcherProxyService} when privileges are updated.
- */
-public class AuthorizationCacheInvalidationTest extends AuthorizationTestBase {
```
removing this cache test makes sense but do we have test for the new cache we added ? I didnt see any unit test for them those caches.

Nishith replied:
The cache is only in containers now, there is a test for it in RemotePrivilegesTest. The extensions will have cache so I will add tests for that under the extension.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,193 @@
+      .build(new CacheLoader<AuthorizationPrivilege, Boolean>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Boolean load(AuthorizationPrivilege authorizationPrivilege) throws Exception {
```
So now there is no auto-refresh of cached entries from the container anymore? This means every TTL seconds, the container data operations need to take penalty?

Nishith replied:
Yes.

chtyim replied:
Is this desirable? How many privilege a contain might have?

Commit:
Addressed comments

Commit:
Fixed a rebase issue

Commit:
Added a delegating Privilege manager.

Nishith commented:
Addressed comments, please take another look. I am adding more unit tests but that will be a separate PR. Also, changed the default cache TTL to 5 mins. After doing some performance tests, if I see that it is not satisfactory then I will add a mechanism to refresh the container cache but that will also be a separate PR, if needed.

Commit:
Addressed unit test comments

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java:
```diff
@@ -78,9 +99,9 @@ public void grant(HttpRequest request, HttpResponder responder) throws Exception
-    LOG.trace("Granting {} on {} to {}", actions, entityId, principal);
-    authorizer.grant(entityId, principal, actions);
-    LOG.debug("Granted {} on {} to {} successfully", actions, entityId, principal);
+    LOG.debug("Granting {} on {} to {}", actions, entityId, principal);
```
Maybe this as trace as well, to keep it consistent with the revoke method.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2744,6 +2730,19 @@
 
   <property>
+    <name>security.authorization.propagate.privileges</name>
+    <value>true</value>
```
Should we have the default as `false`, since no propagation is the more correct behavior? For backward compatibility, user can override this to `true`.

Nishith replied:
I think it might be better to leave it as true for now. For one, it is not working right now because of https://issues.cask.co/browse/CDAP-11497. We can't grant privileges to some system datasets like app.meta. Also, this seems like a major behavior change in a minor release.

chtyim commented on cdap-proto/src/main/java/co/cask/cdap/proto/security/AuthorizationRequest.java:
```diff
@@ -19,6 +19,7 @@
 import co.cask.cdap.api.annotation.Beta;
 import co.cask.cdap.proto.id.EntityId;
 
+import java.util.Arrays;
```
Unused import addition?

chtyim commented on cdap-security/src/main/java/co/cask/cdap/proto/security/AuthorizationPrivilege.java:
```diff
@@ -0,0 +1,79 @@
+ * Key for caching Privileges on containers. This represents a specific privilege on which authorization can be
+ * enforced. The cache stores whether the enforce succeeded or failed.
+ */
+public class AuthorizationPrivilege {
```
You can just extends from `Privilege` and add the `Principal` field?

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,98 @@
+    for (Action action : actions) {
+      try {
+        enforce(entity, principal, action);
+      } catch (UnauthorizedException e) {
```
Shall we preserve the original exception as well by using `addSuppressed`?

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:
```diff
@@ -0,0 +1,123 @@
+  private final RemoteClient remoteClient;
+  private final boolean cacheEnabled;
+
+  private static LoadingCache<AuthorizationPrivilege, Boolean> authPolicyCache;
```
Make it non-static.

chtyim replied:
and final.

chtyim commented:
Just couple comments left

Commit:
Addressed comments

Nishith commented:
@chtyim addressed comments, please take a look.

Commit:
Removed unused imports

chtyim commented:
LGTM

sreevatsanraman commented:
Build has passed. Merging.
